core/include/kernel/mutex.h

```diff
@@ -95,10 +95,20 @@ void condvar_broadcast_debug(struct condvar *cv, const char *fname, int lineno);
 void condvar_wait_debug(struct condvar *cv, struct mutex *m,
 			const char *fname, int lineno);
 #define condvar_wait(cv, m) condvar_wait_debug((cv), (m), __FILE__, __LINE__)
+
+/* Always return TEE_SUCCESS when @timeout_ms is 0 */
+TEE_Result condvar_wait_timeout_debug(struct condvar *cv, struct mutex *m,
+				      uint32_t timeout_ms, const char *fname,
+				      int lineno);
+#define condvar_wait_timeout(cv, m, timeout_ms) \
+	condvar_wait_timeout_debug((cv), (m), (timeout_ms), __FILE__, __LINE__)
 #else
 void condvar_signal(struct condvar *cv);
 void condvar_broadcast(struct condvar *cv);
 void condvar_wait(struct condvar *cv, struct mutex *m);
+/* Always return TEE_SUCCESS when @timeout_ms is 0 */
```

Comment:
I'm sorry, but this isn't true.

It will return `TEE_ERROR_TIMEOUT` If the normal world returns before the condvar has been signaled, regardless of the `timeout_ms` parameter. The normal world is not expected to return early, but we don't trust the normal world so it may very well happen.

This is not as bad as it sounds since callers  of `condvar_wait()` and `condvar_wait_timeout()` should something like the following pattern:
``` C
while (some_condition)
    condvar_wait(...);
```
Caring about the return value is optional since it's the evaluation of `some_condition` that matters. With `condvar_wait_timeout()` you have the option of possibly error out early, but if you care about passed time you need to check the secure time and not trust the normal world.

Reply:
> It will return `TEE_ERROR_TIMEOUT` If the normal world returns before the condvar has been signaled, regardless of the `timeout_ms` parameter. The normal world is not expected to return early, but we don't trust the normal world so it may very well happen.

Oh, that's true. I didn't consider the case `The normal world returns early`. I will remove the comment.

Reply:
Thanks @jenswi-linaro for the feedback.
I think however that a bit of description of how the function behaves and what it returns would help readers/developers.

Reply:
Agreed.

Reply:
Agreed, too.
Could I add the ` Return TEE_ERROR_TIMEOUT If the normal world returns before the condvar has been signaled` as the inline comment for these two functions?

Reply:
Yeah, that's a good clarification.